### PR TITLE
F# - Auto ML samples

### PR DESCRIPTION
**Pull request by kevmal**
Addresses #454

### CODE REVIEW

**Comment by CESARDELATORRE** [comment]
@justinormont - Is this PR about F# samples for AutoML API tested and ready to be merged?

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+            match args with 
+            | [|_; "svg"|] -> 
+                pl.sdev("svg")
+                "TaxiRegressionDistribution.svg"
+            | _ -> 
+                pl.sdev("pngcairo")
+                "TaxiRegressionDistribution.png"
+        pl.sfnam chartFileName
+
+        // use white background with black foreground
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]:
```suggestion
        // Use white background with black foreground
```

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        let yMaxLimit = 35.0  // Rides larger than $35 are not shown in the chart
+        pl.env(xMinLimit, xMaxLimit, yMinLimit, yMaxLimit, AxesScale.Independent, AxisBox.BoxTicksLabelsAxes)
+
+        // Set scaling for mail title text 125% size of default
+        pl.schr(0.0, 1.25)
+
+        // The main title
+        pl.lab("Measured", "Predicted", "Distribution of Taxi Fare Prediction")
+
+        // plot using different colors
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]:
```suggestion
        // Plot using different colors
```

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        // see http://plplot.sourceforge.net/examples.php?demo=02 for palette indices
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]:
```suggestion
        // See: http://plplot.sourceforge.net/examples.php?demo=02 for palette indices
```

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            readTaxiTripCsv testDataPath
+            |> Seq.truncate numberOfRecordsToRead
+            |> Seq.toArray
+
+        let totalNumber = double testData.Length
+
+        // This code is the symbol to paint
+        let code = char 9
+
+        // plot using other color
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]:
```suggestion
        // Plot using other color
```

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                    printfn "-------------------------------------------------"
+                    printfn "Predicted : %f" y
+                    printfn "Actual:    %f" x
+                    printfn "-------------------------------------------------"
+                    xTotal + x, yTotal + y,  xyMultiTotal + x*y, xSquareTotal + x*x
+                )
+     
+        // Regression Line calculation explanation:
+        // https://www.khanacademy.org/math/statistics-probability/describing-relationships-quantitative-data/more-on-regression/v/regression-line-example
+
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]:
```suggestion
```

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        // Set the metric that AutoML will try to optimize over the course of the experiment.
+        experimentSettings.OptimizingMetric <- RegressionMetric.RootMeanSquaredError
+
+        // Set the cache directory to null.
+        // This will cause all models produced by AutoML to be kept in memory 
+        // instead of written to disk after each run, as AutoML is training.
+        // (Please note: for an experiment on a large dataset, opting to keep all 
+        // models trained by AutoML in memory could cause your system to run out 
+        // of memory.)
+        experimentSettings.CacheDirectory <- null
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]: Setting to `null` is generally a bad idea. Not sure why in the C# version we first rather warn to not set to null, and then set to null. Mainly this limits the number of iterations a user can run.

Comment (justinormont) [reply samples/fsharp/getting-started/AdvancedExperiment_AutoML/AdvancedTaxiFarePrediction/Program.fs]:
```suggestion
        //experimentSettings.CacheDirectory <- null
```

File: samples/fsharp/getting-started/AdvancedExperiment_AutoML/README.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+// Set the metric that AutoML will try to optimize over the course of the experiment.
+experimentSettings.OptimizingMetric <- RegressionMetric.RootMeanSquaredError
+
+// Set the cache directory to null.
+// This will cause all models produced by AutoML to be kept in memory 
+// instead of written to disk after each run, as AutoML is training.
+// (Please note: for an experiment on a large dataset, opting to keep all 
+// models trained by AutoML in memory could cause your system to run out 
+// of memory.)
+experimentSettings.CacheDirectory <- null
```
Comment (justinormont) [inline samples/fsharp/getting-started/AdvancedExperiment_AutoML/README.md]:
```suggestion
//experimentSettings.CacheDirectory <- null
```

File: samples/fsharp/getting-started/BinaryClassification_AutoML/SentimentAnalysis/Program.fs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+// Print top models found by AutoML
+printfn ""
+printfn "Top models ranked by accuracy --"
+experimentResult.RunDetails
+|> Seq.filter (fun r -> not (isNull r.ValidationMetrics) && not (Double.IsNaN r.ValidationMetrics.Accuracy))
+|> Seq.sortBy (fun x -> x.ValidationMetrics.Accuracy)
+|> Seq.truncate 3
+|> Seq.iteri (fun i x -> ConsoleHelper.printBinaryIterationMetrics (i + 1) x.TrainerName x.ValidationMetrics x.RuntimeInSeconds) 
+
+
```
Comment (justinormont) [inline samples/fsharp/getting-started/BinaryClassification_AutoML/SentimentAnalysis/Program.fs]:
```suggestion
```

File: samples/fsharp/getting-started/Regression_AutoML/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+let testDataView = mlContext.Data.LoadFromTextFile<TaxiTrip>(testDataPath, hasHeader = true, separatorChar = ',')
+```
+
+## Step 2: Building a Machine Learning Model using AutoML
+
+Instantiate and run an AutoML experiment. In doing so, specify how long the experiment should run in seconds (`ExperimentTime`), and set a progress handler that will receive notifications after AutoML trains & evaluates each new model.
+
+```fsharp
+// Run AutoML binary classification experiment
+let experimentResult = mlContext.Auto() .CreateRegressionExperiment(experimentTimeInSeconds).Execute(trainingDataView, labelColumnName, progressHandler= progressHandler)
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/README.md]:
```suggestion
let experimentResult = mlContext.Auto().CreateRegressionExperiment(experimentTimeInSeconds).Execute(trainingDataView, labelColumnName, progressHandler= progressHandler)
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        let yMaxLimit = 35.0  // Rides larger than $35 are not shown in the chart
+        pl.env(xMinLimit, xMaxLimit, yMinLimit, yMaxLimit, AxesScale.Independent, AxisBox.BoxTicksLabelsAxes)
+
+        // Set scaling for mail title text 125% size of default
+        pl.schr(0.0, 1.25)
+
+        // The main title
+        pl.lab("Measured", "Predicted", "Distribution of Taxi Fare Prediction")
+
+        // plot using different colors
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
        // Plot using different colors
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        // see http://plplot.sourceforge.net/examples.php?demo=02 for palette indices
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
        // See: http://plplot.sourceforge.net/examples.php?demo=02 for palette indices
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+            readTaxiTripCsv testDataPath
+            |> Seq.truncate numberOfRecordsToRead
+            |> Seq.toArray
+
+        let totalNumber = double testData.Length
+
+        // This code is the symbol to paint
+        let code = char 9
+
+        // plot using other color
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
        // Plot using other color
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        //pl.col0(9) //Light Green
+        //pl.col0(4) //Red
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
        //pl.col0(4) // Red
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        pl.col0(2) //Blue
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
        pl.col0(2) // Blue
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+                    printfn "-------------------------------------------------"
+                    printfn "Predicted : %f" y
+                    printfn "Actual:    %f" x
+                    printfn "-------------------------------------------------"
+                    xTotal + x, yTotal + y,  xyMultiTotal + x*y, xSquareTotal + x*x
+                )
+     
+        // Regression Line calculation explanation:
+        // https://www.khanacademy.org/math/statistics-probability/describing-relationships-quantitative-data/more-on-regression/v/regression-line-example
+
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+// STEP 6: Save/persist the trained model to a .ZIP file
+mlContext.Model.Save(trainedModel, trainingDataView.Schema, modelPath)
+
+printfn "The model is saved to %s" modelPath
+
+
+// Make a single test prediction loading the model from .ZIP file
+
+//Sample:
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
// Sample: 
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+// STEP 6: Save/persist the trained model to a .ZIP file
+mlContext.Model.Save(trainedModel, trainingDataView.Schema, modelPath)
+
+printfn "The model is saved to %s" modelPath
+
+
+// Make a single test prediction loading the model from .ZIP file
+
+//Sample: 
+//vendor_id,rate_code,passenger_count,trip_time_in_secs,trip_distance,payment_type,fare_amount
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
// vendor_id,rate_code,passenger_count,trip_time_in_secs,trip_distance,payment_type,fare_amount
```

File: samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+//VTS,1,1,1140,3.75,CRD,15.5
```
Comment (justinormont) [inline samples/fsharp/getting-started/Regression_AutoML/TaxiFarePrediction/Program.fs]:
```suggestion
// VTS,1,1,1140,3.75,CRD,15.5
```

**Comment by justinormont** [comment]
> @justinormont - Is this PR about F# samples for AutoML API tested and ready to be merged?

@CESARDELATORRE: It looks good and is a conversion of the existing C# sample to F#; I haven't tested it though.

**Comment by kevmal** [comment]
Is this good to go? I could post the console output of the samples if that helps?

**Comment by bamurtaugh** [comment]
@CESARDELATORRE All 4 AutoML samples worked when I tested them and should be ready to merge.

**Pull request merged**